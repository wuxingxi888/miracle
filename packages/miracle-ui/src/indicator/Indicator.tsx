import { defineComponent, type ExtractPropTypes } from 'vue';

import { createNamespace, makeNumericProp, makeStringProp } from '../utils';

const [name, bem] = createNamespace('indicator');

export const indicatorProps = {
    type: makeStringProp<IndicatorType>('circle'),
    color: String,
    direction: makeStringProp<IndicatorDirection>('horizontal'),
    size: makeNumericProp(1),
    active: makeNumericProp(0),
};

export type IndicatorProps = ExtractPropTypes<typeof indicatorProps>;

export type IndicatorDirection = 'horizontal' | 'vertical';

export type IndicatorType = 'circle' | 'square' | 'line';

export default defineComponent({
    name,

    props: indicatorProps,

    setup(props) {
        return () => {
            const { type, color, size, active, direction } = props;

            const renderIndicator = (shape: string, isActive: boolean) => {
                const style = isActive ? { backgroundColor: color } : undefined;
                const classes = bem(shape, {
                    active: isActive,
                    [direction]: shape === 'line',
                });

                return <i style={style} class={classes}></i>;
            };

            const renderIndicators = (index: number) => {
                const isActive = index === active;
                return renderIndicator(type, isActive);
            };

            return (
                <div class={bem()}>
                    <div class={bem('indicators', { [direction]: true })}>
                        {Array(size)
                            .fill('')
                            .map((_, index) => renderIndicators(index))}
                    </div>
                </div>
            );
        };
    },
});
