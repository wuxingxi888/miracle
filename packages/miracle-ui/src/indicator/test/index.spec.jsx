import { mount } from '../../../test';
import Indicator from '../../indicator';

test('should allow to custom Indicator type', () => {
  const wrapper = mount({
    render: () => <Indicator type="square" size="3" active="1" />,
  });

  expect(wrapper.find('.mi-indicator').html()).toMatchSnapshot();
});
