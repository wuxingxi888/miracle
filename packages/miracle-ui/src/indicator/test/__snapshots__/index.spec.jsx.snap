// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should allow to custom Indicator type 1`] = `
<div class="mi-indicator">
  <div class="mi-indicator__indicators mi-indicator__indicators--horizontal">
    <i class="mi-indicator__square">
    </i>
  </div>
</div>
`;
